src/core/string_consumer.cpp

```diff
@@ -0,0 +1,196 @@
+
+#include "../string_func.h"
+
+#if defined(STRGEN) || defined(SETTINGSGEN)
+#include "../error_func.h"
+#else
+#include "../debug.h"
+#endif
+
+#include "../safeguards.h"
+
+const std::string_view StringConsumer::WHITESPACE_NO_NEWLINE = "\t\v\f\r ";
+const std::string_view StringConsumer::WHITESPACE_OR_NEWLINE = "\t\n\v\f\r ";
+
+void StringConsumer::LogError(std::string &&msg)
+{
+#if defined(STRGEN) || defined(SETTINGSGEN)
+	FatalErrorI(std::move(msg));
+#else
+	DebugPrint("misc", 0, std::move(msg));
+#endif
+}
+
+std::optional<uint8_t> StringConsumer::PeekUint8() const
+{
+	if (this->GetBytesLeft() < 1) return std::nullopt;
+	return static_cast<uint8_t>(this->src[this->position]);
+}
+
+std::optional<uint16_t> StringConsumer::PeekUint16LE() const
+{
+	if (this->GetBytesLeft() < 2) return std::nullopt;
+	return static_cast<uint8_t>(this->src[this->position]) |
+		static_cast<uint8_t>(this->src[this->position + 1]) << 8;
+}
+
+std::optional<uint32_t> StringConsumer::PeekUint32LE() const
+{
+	if (this->GetBytesLeft() < 4) return std::nullopt;
+	return static_cast<uint8_t>(this->src[this->position]) |
+		static_cast<uint8_t>(this->src[this->position + 1]) << 8 |
+		static_cast<uint8_t>(this->src[this->position + 2]) << 16 |
+		static_cast<uint8_t>(this->src[this->position + 3]) << 24;
+}
+
+std::optional<uint64_t> StringConsumer::PeekUint64LE() const
+{
+	if (this->GetBytesLeft() < 8) return std::nullopt;
+	return static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position])) |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 1])) << 8 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 2])) << 16 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 3])) << 24 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 4])) << 32 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 5])) << 40 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 6])) << 48 |
+		static_cast<uint64_t>(static_cast<uint8_t>(this->src[this->position + 7])) << 56;
+}
+
+std::optional<char> StringConsumer::PeekChar() const
+{
+	auto result = this->PeekUint8();
+	if (!result.has_value()) return {};
+	return static_cast<char>(*result);
+}
+
+std::pair<StringConsumer::size_type, char32_t> StringConsumer::PeekUtf8() const
+{
+	auto buf = this->src.substr(this->position);
+	return DecodeUtf8(buf);
+}
+
+std::string_view StringConsumer::Peek(size_type len) const
+{
+	auto buf = this->src.substr(this->position);
+	if (len == std::string_view::npos) {
+		len = buf.size();
+	} else if (len > buf.size()) {
+		len = buf.size();
+	}
+	return buf.substr(0, len);
+}
+
+void StringConsumer::Skip(size_type len)
+{
+	if (len == std::string_view::npos) {
+		this->position = this->src.size();
+	} else if (size_type max_len = GetBytesLeft(); len > max_len) {
+		this->LogError(fmt::format("Source buffer too short: {} > {}", len, max_len));
```

Comment:
```suggestion
		LogError(fmt::format("Source buffer too short: {} > {}", len, max_len));
```
Although `StringConsumer::LogError` might be more correct, but calling a `static` function with `this->` is definitely weird.